test_plans/is_device_copyable.asciidoc

```diff
@@ -0,0 +1,81 @@
+
+* The application defines the trait `is_device_copyable_v<T>` to `true`;
+* Type `T` has at least one eligible copy constructor, move constructor, copy assignment operator, or move assignment operator;
+* Each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is public;
+* When doing an inter-device transfer of an object of type `T`, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;
+* Type `T` has a public non-deleted destructor;
+* The destructor has no effect when executed on the device.
+
+Define a new class of type `T` with public default constructor, destructor, copy constructor, copy assignment operator and several public members:
+```
+class T {
+  public:
+    T() = default;
+    ~T() = default;
+    T(T& t) : a(t.a), b(t.b), c(t.c) {}
+
+    T& operator=(const T& t) {
+        a = t.a;
+        b = t.b;
+        c = t.c;
+    }
+
+    int a;
+    float b;
+    char c;
+};
+```
+
+== Tests
+
+=== Cpp17UnaryTrait requirements.
+
+`is_device_copyable` must meet the https://en.cppreference.com/w/cpp/named_req/UnaryTypeTrait[Cpp17UnaryTrait] requirements:
+
+    * DefaultConstructible and CopyConstructible. Verify that:
+
+        ** `std::is_default_constructible_v<is_device_copyable<T>> == true`
+        ** `std::is_copy_constructible_v<is_device_copyable<T>> == true`
+
+    * Takes one template type parameter (additional template parameters are optional and allowed). Verify that: `std::is_constructible_v<is_device_copyable<T>> == true`
+
+    * Publicly and unambiguously derived from a specialization of integral_constant, known as its base characteristic. Verify that +
+    `std::is_base_of_v<std::integral_constant<bool, true>, is_device_copyable<T>> == true` or
+    `std::is_base_of_v<std::integral_constant<bool, false>, is_device_copyable<T>> == true`.
+
+    * The member names of the base characteristic (`std::integral_constant`) are not hidden and are unambiguously available. Verify that:
+
+        ** `std::integral_constant<bool, true>::value_type` and `is_device_copyable<T>::value_type` are the same type
+
+        ** `std::integral_constant<bool, true>::type` and `is_device_copyable<T>::type` are the same type
+        ** Operators `std::integral_constant<bool, true>::value_type() and `is_device_copyable<T>::value_type()` have the same return type
+        ** `std::integral_constant<bool, true>::operator()` returns true
+
+=== Specialization for user defined class
+
+Add a specialization for `T`:
+
+`template<> +
+struct is_device_copyable<T> : std::true_type {}` +
+
+* Verify that `is_device_copyable<T>::value == true`
+* Create an instance of class `T` on the host and pass an accessor to it in device code;
```

Comment:
Should we add a test here:

* Verify that the object of type `T` in the kernel has the same value that was initialized on the host.

Reply:
I planned to initialize new object on device with another data and pass this data to the host object, so main verification would be the last point of this section.

Reply:
Yes, I see that.  The last bullet point verifies that the object of type `T` can be copied from device-to-host.  The bullet I proposed checks the opposite case: that the object can be copied from host-to-device.